src/tiered/tiered_work.c

```diff
@@ -0,0 +1,151 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+/*
+ * __wt_tiered_push_work --
+ *     Push a work unit to the queue. Assumes it is passed an already filled out structure.
+ */
+void
+__wt_tiered_push_work(WT_SESSION_IMPL *session, WT_TIERED_WORK_UNIT *entry)
+{
+    WT_CONNECTION_IMPL *conn;
+
+    conn = S2C(session);
+    __wt_spin_lock(session, &conn->tiered_lock);
+    TAILQ_INSERT_TAIL(&conn->tieredqh, entry, q);
+    WT_STAT_CONN_INCR(session, tiered_work_units_created);
+    __wt_spin_unlock(session, &conn->tiered_lock);
+    __wt_cond_signal(session, conn->tiered_cond);
+    return;
+}
+
+/*
+ * __wt_tiered_pop_work --
+ *     Pop a work unit of the given type from the queue. If a maximum value is given, only return a
+ *     work unit that is less than the maximum value. The caller is responsible for freeing the
+ *     returned work unit structure.
+ */
+void
+__wt_tiered_pop_work(
+  WT_SESSION_IMPL *session, uint32_t type, uint64_t maxval, WT_TIERED_WORK_UNIT **entryp)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_TIERED_WORK_UNIT *entry;
+
+    *entryp = entry = NULL;
+
+    conn = S2C(session);
+    if (TAILQ_EMPTY(&conn->tieredqh))
+        return;
+    __wt_spin_lock(session, &conn->tiered_lock);
+
+    TAILQ_FOREACH (entry, &conn->tieredqh, q) {
+        if (FLD_ISSET(type, entry->type) && (maxval == 0 || entry->op_val < maxval)) {
+            TAILQ_REMOVE(&conn->tieredqh, entry, q);
+            WT_STAT_CONN_INCR(session, tiered_work_units_done);
```

Comment:
A technicality, but is work done yet? (statistic name implies it is done).

Reply:
No, it really is counting work units dequeued. Since the caller is responsible for doing work and freeing the entry, should we assume all work functions increment a stat (and maybe have one stat per work unit type) or should I rename this to `dequeued` instead of `done`?

Reply:
I renamed it to `dequeued`. We can always change it later if we want.